beacon_node/beacon_chain/src/validator_custody.rs

```diff
@@ -0,0 +1,433 @@
+        }
+
+        // Completed registrations, now check if the validator custody requirement has changed
+        let Some(new_validator_custody) = registrations.latest_validator_custody_requirement()
+        else {
+            return;
+        };
+
+        let current_cgc = self.custody_group_count(spec);
+        let validator_custody_count_at_head =
+            self.validator_custody_count_at_head.load(Ordering::Relaxed);
+
+        if new_validator_custody != validator_custody_count_at_head {
+            tracing::debug!(
+                old_count = validator_custody_count_at_head,
+                new_count = new_validator_custody,
+                "Validator count at head updated"
+            );
+            self.validator_custody_count_at_head
+                .store(new_validator_custody, Ordering::Relaxed);
+
+            let updated_cgc = self.custody_group_count(spec);
+            // Send the message to network only if there are more columns subnets to subscribe to
+            if updated_cgc > current_cgc {
+                tracing::debug!(
+                    old_cgc = current_cgc,
+                    updated_cgc,
+                    "Custody group count updated"
+                );
+                if let Err(e) = self
+                    .sender
+                    .send(CustodyContextMessage::HeadCustodyCountChanged {
+                        new_custody_count: updated_cgc,
+                    })
+                {
+                    tracing::error!(error=?e, "Failed to send custody context message");
+                }
+            }
+        }
+    }
+
+    /// The custody count that we advertise to our peers in our metadata and
+    /// enr values.
+    pub fn advertised_custody_group_count(&self, spec: &ChainSpec) -> u64 {
+        if self.persisted_is_supernode {
+            return spec.number_of_custody_groups;
+        }
+        let advertised_validator_custody = self
+            .advertised_validator_custody_count
+            .load(Ordering::Relaxed);
+
+        // If there are no validators, return the minimum custody_requirement
+        if advertised_validator_custody > 0 {
+            advertised_validator_custody
+        } else {
+            spec.custody_requirement
+        }
+    }
+
+    /// The custody count that we use to custody columns currently.
+    ///
+    /// This function should be called when figuring out how many columns we
+    /// need to custody when receiving blocks over gossip/rpc or during sync.
+    pub fn custody_group_count(&self, spec: &ChainSpec) -> u64 {
+        if self.current_is_supernode {
+            return spec.number_of_custody_groups;
+        }
+        let validator_custody_count_at_head =
+            self.validator_custody_count_at_head.load(Ordering::Relaxed);
+
+        // If there are no validators, return the minimum custody_requirement
+        if validator_custody_count_at_head > 0 {
```

Comment:
Could this be simplified to
`max(validator_custody_count_at_head, spec.custody_requirement)`

Reply:
Yeah originally I was also concerned about the case where the value is > 0 but less than `spec.custody_requirement`, but this is ensured here, so no strong opinion to keep it as it is.
https://github.com/sigp/lighthouse/blob/b23a4dcc787728109893b0e09520a8fde15d1c8f/beacon_node/beacon_chain/src/validator_custody.rs#L93-L98